lib/Analysis/Membar.cpp

```diff
@@ -64,29 +60,74 @@ void MembarAnalysis::resolve(FunctionOpInterface funcOp,
     // so overwrite the output state entirely.
     outputBlockInfoMap[block] = inputBlockInfo;
     // Update the successors
-    for (auto *successor : successors) {
+    for (VirtualBlock successor : successors) {
       inputBlockInfoMap[successor].join(outputBlockInfoMap[block]);
       blockList.emplace_back(successor);
     }
   }
 
   // Update the final dangling buffers that haven't been synced
-  auto &funcBlockInfo = (*funcBlockInfoMap)[funcOp];
-  funcOp.walk<WalkOrder::PreOrder>([&](Block *block) {
-    block->walk([&](triton::ReturnOp returnOp) {
-      funcBlockInfo.join(outputBlockInfoMap[block]);
+  BlockInfo &funcBlockInfo = (*funcBlockInfoMap)[funcOp];
+  funcOp.walk<WalkOrder::PreOrder>([&](triton::ReturnOp returnOp) {
+    // Find the virtual block that contains this op.
+    SmallVector<std::pair<VirtualBlock, BlockInfo>> virtualBlocks;
+    for (auto &[block, blockInfo] : outputBlockInfoMap) {
+      if (block.first == returnOp->getBlock())
+        virtualBlocks.emplace_back(block, blockInfo);
+    }
+    // The last virtual block in this group contains the terminator.
+    auto maxIt = llvm::max_element(virtualBlocks, [&](auto &lhs, auto &rhs) {
+      assert(lhs.first.first == rhs.first.first);
+      Block::iterator lhsIt = lhs.first.second, rhsIt = rhs.first.second;
+      return !lhsIt.isValid() ||
+             (rhsIt.isValid() && lhsIt->isBeforeInBlock(&*rhsIt));
     });
+
+    funcBlockInfo.join(maxIt->second);
   });
 }
 
 void MembarAnalysis::visitTerminator(Operation *op,
-                                     SmallVector<Block *> &successors) {
-  if (auto branchInterface = dyn_cast<BranchOpInterface>(op)) {
-    Block *parentBlock = branchInterface->getBlock();
-    successors.append(std::begin(parentBlock->getSuccessors()),
-                      std::end(parentBlock->getSuccessors()));
+                                     SmallVector<VirtualBlock> &successors) {
+  if (isa<BranchOpInterface>(op)) {
+    for (Block *successor : op->getSuccessors())
+      successors.emplace_back(successor, Block::iterator());
     return;
   }
+
+  if (auto br = dyn_cast<RegionBranchOpInterface>(op)) {
+    SmallVector<RegionSuccessor> regions;
+    br.getSuccessorRegions(RegionBranchPoint::parent(), regions);
```

Comment:
Why the point is `parent`?

Reply:
This is getting region successors originating from the parent op itself. I.e. an `scf.for` can branch to itself or its body region. Its body region can branch back to the parent or itself.

Reply:
OK I get it. This is the point that gets into the single region of `scf.for` for example or gets back to `scf.for` operation itself.

The code below handles the terminator in `scf.for`.